packages/flutter_tools/lib/src/base/process.dart

```diff
@@ -542,6 +542,9 @@ class _DefaultProcessUtils implements ProcessUtils {
     } on Exception catch (error) {
       _logger.printTrace('$cli failed with $error');
       return false;
+    } on ArgumentError catch (error) {
```

Comment:
Why are we even calling it in this case?

Reply:
Sometimes we call `exitsHappy` to verify if the binary exists, e.g. https://github.com/flutter/flutter/blob/071d4eb2c8cb20f2f28c0740046ecfae69e5c7ae/packages/flutter_tools/lib/src/ios/mac.dart#L38

Reply:
Ahhh ok, I see now, and we're catching `ArgumentError` from `_processManager`.  Is this more reliable than just calling https://api.flutter.dev/flutter/package-process_process/ProcessManager/canRun.html?

Reply:
Not more reliable, but I think also not less reliable(?) The goal with `processUtils` was to have a small, concise API surface.

Reply:
Yes, I mean we could call that in here first and return false if it's false, rather than trying to catch an argument exception.  That way if some other code starts throwing that we don't end up swallowing the unrelated exception.

Reply:
i.e. on line 541 of this file (or maybe line 539), `if (!_processManager.canRun(cli.first), environment: environment)) { return false; }` - probably also with either asserts or guards against `cli` being null or empty.

Reply:
Oh, I see what you mean

Reply:
That would be changing what this function does slightly, however, by discarding the arguments.

Reply:
`DeviceManager.canListAnything` should probably check `discoverer.supportsPlatform` before querying `discoverer.canListAnything`.  The similar code in `EmulatorManager` in emulator.dart probably needs a similar fix. (Sorry for not suggesting this earlier, but the fix in this PR was needed as well.)

Reply:
RE discarding arguments - after checking `canRun`, you can still use the existing logic, but avoid trying to catch an argument error.  

My concern with catching argument error is that it can be thrown for a number of reasons, including errors in the upstream code (e.g. someone accidentally makes the string null, or passes in an empty array).